crates/bevy_render/src/batching/mod.rs

```diff
@@ -1,87 +1,89 @@
+use bevy_asset::AssetId;
 use bevy_ecs::component::Component;
 use nonmax::NonMaxU32;
 
-use crate::render_phase::{CachedRenderPipelinePhaseItem, PhaseItem, RenderPhase};
+use crate::{
+    mesh::Mesh,
+    render_phase::{CachedRenderPipelinePhaseItem, DrawFunctionId, RenderPhase},
+    render_resource::CachedRenderPipelineId,
+};
 
 /// Add this component to mesh entities to disable automatic batching
 #[derive(Component)]
 pub struct NoAutomaticBatching;
 
-pub trait BatchMeta<T: BatchMeta<T>> {
-    fn matches(&self, other: &T) -> bool;
+/// Data necessary to be equal for two draw commands to be mergeable
+///
+/// This is based on the following assumptions:
+/// - Only entities with prepared assets (pipelines, materials, meshes) are
+///   queued to phases
+/// - View bindings are constant across a phase for a given draw function as
+///   phases are per-view
+/// - `prepare_mesh_uniforms` is the only system that performs this batching
+///   and has sole responsibility for preparing the per-object data. As such
+///   the mesh binding and dynamic offsets are assumed to only be variable as a
+///   result of the `prepare_mesh_uniforms` system, e.g. due to having to split
+///   data across separate uniform bindings within the same buffer due to the
+///   maximum uniform buffer binding size.
+pub struct BatchMeta<T: PartialEq> {
+    /// The pipeline id encompasses all pipeline configuration including vertex
+    /// buffers and layouts, shaders and their specializations, bind group
+    /// layouts, etc.
+    pub pipeline_id: CachedRenderPipelineId,
+    /// The draw function id defines the RenderCommands that are called to
+    /// set the pipeline and bindings, and make the draw command
+    pub draw_function_id: DrawFunctionId,
+    pub material_bind_group_id: Option<T>,
+    pub mesh_asset_id: AssetId<Mesh>,
+    pub dynamic_offset: Option<NonMaxU32>,
 }
 
-struct BatchState<T: BatchMeta<T>> {
-    meta: Option<T>,
-    /// The base index in the object data binding's array
-    index: Option<NonMaxU32>,
-    /// The dynamic offset of the data binding
-    dynamic_offset: Option<NonMaxU32>,
-    /// The number of entities in the batch
-    count: u32,
-    item_index: usize,
-}
-
-impl<T: BatchMeta<T>> Default for BatchState<T> {
-    fn default() -> Self {
-        Self {
-            meta: Default::default(),
-            index: Default::default(),
-            dynamic_offset: Default::default(),
-            count: Default::default(),
-            item_index: Default::default(),
-        }
+impl<T: PartialEq> PartialEq for BatchMeta<T> {
```

Comment:
don't think you need the bound on T ... in fact i dn't think you need T at all

Reply:
If I don't add the bound on T then it complains that it doesn't implement PartialEq. And I could maybe be convinced to have the same component for 2D / 3D MaterialBindGroupId and then it wouldn't be needed.

Reply:
Sorry I missed the T on the bind group id. All good.